iravelmakina/svd-ivelmakina
# Singular Value Decomposition implementation and Recommendation Algorithm

### Description:
This pull request adds functionality for performing Singular Value Decomposition (SVD) on a given matrix, verifying its accuracy, and applying it to a movie recommendation system using movie ratings data. It also includes visualization of users and movies in the latent feature space and a function to recommend movies to users.

### Changes:
- Added `custom_svd` function for SVD computation
- Added `verify_svd` function to check the correctness of SVD
- Loaded and preprocessed movie ratings data
- Implemented dimensionality reduction with SVD
- Visualized users and movies in 3D latent feature space
- Created a function to recommend top 10 movies for a given user

### How to Test:
1. Run the `custom_svd` function and verify the output matrices (U, Sigma, V^T).
2. Load the movie ratings data from the specified file path.
3. Preprocess the data by removing users and movies with insufficient ratings.
4. Fill missing values in the ratings matrix with mean ratings.
5. Perform SVD on the demeaned ratings matrix, mind that k defines error and complexity of calculations.
6. Visualize the users and movies in the 3D latent feature space.
7. Use the `recommend_movies` function to generate top 10 movie recommendations for different users and verify the output.

### Screenshots:
1.
 ![image](https://github.com/iravelmakina/svd-ivelmakina/assets/150923968/8f9c9c79-0991-406b-959f-bda4f2e100b4)
2-6. 
![image](https://github.com/iravelmakina/svd-ivelmakina/assets/150923968/34199c7e-aa99-4147-ab58-58c5cedb4092)

![image](https://github.com/iravelmakina/svd-ivelmakina/assets/150923968/b114630c-f4d4-4dcb-83e8-545191512267)
![image](https://github.com/iravelmakina/svd-ivelmakina/assets/150923968/06e3c160-7055-452c-95c2-f84f178a5629)
7. 
![image](https://github.com/iravelmakina/svd-ivelmakina/assets/150923968/1e92923f-e1e9-47ee-9b76-da93049098ac)
![image](https://github.com/iravelmakina/svd-ivelmakina/assets/150923968/643dc782-4b82-41d8-8c06-b8fa6c8ffff0)
![image](https://github.com/iravelmakina/svd-ivelmakina/assets/150923968/af087f44-acf9-42a0-97dc-6a39300c5a6a)
